### PR TITLE
DRAFT: Git Bisect Post 2020-04

### PR DESCRIPTION
**Pull request by shivarajagopal**
Signed-off-by: Shiva Rajagopal <shiva.rajagopal@outlook.com>

### CODE REVIEW

File: _drafts/git-bisect.md [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+Bisecting: 6 revisions left to test after this (roughly 3 steps)
+[26338040b292e110c2086de6bc42d1c232005a58] Refactor python script with classes
+```
+
+In a very helpful output, git is telling us how many different commits we still
+need to run through, and is counting down about how many test steps we have
+left. The view of our bisect looks something like this right now:
+
+```
+RED
```
Comment (shivarajagopal) [inline _drafts/git-bisect.md:219]: Couldn't figure out how to do color in Markdown, but any suggestions are welcome. For now, I just labeled the color before what I want the next set of commits to be.

File: _drafts/git-bisect.md [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+pace, with all of its contributors committing refactors, improvements, and
+bugfixes every day. You’ve been using git commands so much recently that you
+can rely on your fingers’s muscle memory to do all of your git operations for
+you. All of a sudden, you get that dreaded message from a teammate:
+
+> The firmware flasher is broken, and we don’t know when it broke
+
+If you’re anything like me, your reaction looks something like this.
+
+| ![](/img/git-bisect/gambino.gif) |
```
Comment (shivarajagopal) [inline _drafts/git-bisect.md:17]: This was the best I could do for a picture with a decent-looking caption, but suggestions are welcome. Maybe center it?

**Comment by franc0is** [comment]
@shivarajagopal FYI I pushed a change to center the figures (with help from @sirlantis)

File: _drafts/git-bisect.md [hunk 3]
```diff
@@ -0,0 +1,600 @@
+---
+title: "Hunting Bugs with Git Bisect"
+description: "An introduction to bisecting a set of changes as a bug-hunting tool, followed by some deeper techniques to speed up the process of tracking down when behavior is introduced"
+author: shiva
+image: /img/git-bisect/git-logo.png
+---
+<style>
```
Comment (franc0is) [inline _drafts/git-bisect.md:7]: Woops I need to nuke this

**Comment by franc0is** [comment]
FYI @shivarajagopal I pushed a change to highlight your "bisect state" properly.

File: _drafts/git-bisect.md [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+Git has now checked out commit 26338040 for us to test. This commit is
+effectively the midpoint between v1.0 and the head of our repo. Let’s run our
+python script to see what’s happening.
+
+```
+>>> python update.py
+Verification failed!
+```
+
+We now know things were still bad at commit 26338040. Let’s tell git about that.
```
Comment (chrisc11) [inline _drafts/git-bisect.md:262]:
```suggestion
We now know things were still bad at commit 26338040. Let’s tell git about that:
```

File: _drafts/git-bisect.md [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+As we saw at the end of the last section, our bug seems to have been introduced
+with commit e0c258c0. Let’s check that changeset for verification changes...
+
+```diff
++ uint8_t readback_data[size];
++ nor_flash_dev->operations.read(sector_to_compare, readback_data, size);
++ int return_code = memcmp((void *) DATA_PAYLOAD_ADDR, readback_data, size);
++ return (return_code == 0)
+```
+
```
Comment (chrisc11) [inline _drafts/git-bisect.md:424]: Maybe: Might + in the function signature as part of the diff so its easier to follow along with your description below. I think that would be something like:

```diff
+// @return 0 if contents match
+ int verify_copy(uint32_t sector_to_compare) {
[...]
+     uint8_t readback_data[size];
+     nor_flash_dev->operations.read(sector_to_compare, readback_data, size);
+     int return_code = memcmp((void *) DATA_PAYLOAD_ADDR, readback_data, size);
+     return (return_code == 0)
+}
```

Comment (shivarajagopal) [reply _drafts/git-bisect.md:421]: Good call, will do.

File: _drafts/git-bisect.md (hunk 3 above)
Comment (shivarajagopal) [reply _drafts/git-bisect.md]: Went ahead and removed it myself in 81febd407.

File: _drafts/git-bisect.md (hunk 5 above)
Comment (shivarajagopal) [reply _drafts/git-bisect.md:423]: Done!

**Comment by shivarajagopal** [comment]
@franc0is @chrisc11 Thanks so much for the feedback and for compensating for my inability to make my graphics look good :)

**Comment by tyhoff** [comment]
@shivarajagopal Post is published and up on Interrupt!

**Comment by shivarajagopal** [comment]
Thanks @tyhoff!

**Pull request merged**